### ISSUE TITLE
Provide a way to set custom DEBUG logger

### ISSUE THREAD

**Issue by ogerardin** [gin-gonic/gin#1877]
- go version: not relevant
- gin version (or commit ref): 1.3.0
- operating system: not relevant

## Description
Gin uses the package function `log.Printf` to log debug information, notably during startup, as follows:

```
func debugPrint(format string, values ...interface{}) {
	if IsDebugging() {
		log.Printf("[GIN-debug] "+format, values...)
	}
}
```
This function `log.Printf` uses the standard logger which writes to stderr. I want to direct Gin's debug log to another logger (native or logrus), but using a static logger prevents me from doing so. It would be easy to allow a custom logger to be passed to Gin.

PS: I know you can plug a middleware logger, but that won't solve my problem, since I want the debug log...

**Comment by thinkerou** [comment]
please see https://github.com/gin-gonic/gin#how-to-write-log-file and https://github.com/gin-gonic/gin#custom-log-format, maybe can help you.

**Comment by ogerardin** [comment]
> please see https://github.com/gin-gonic/gin#how-to-write-log-file and https://github.com/gin-gonic/gin#custom-log-format, maybe can help you.

This does not apply to DEBUG logging which directly calls the log.Printf function.

**Comment by fdingiit** [comment]
Is there a way to de-function logging? Meaning no log at all.

**Comment by sneak** [comment]
https://github.com/gin-gonic/gin/blob/b67bc8f00502b1f555b9db64cd2dfc03098cfc8f/debug.go#L24

It looks like if you set `gin.DebugPrintRouteFunc` to a compatible function (`func(httpMethod, absolutePath, handlerName string, nuHandlers int)`) then it will log the route debug info using that, but you're right, the rest of the debug stuff just writes to `gin.DefaultWriter` which is stdout normally.

Looks like it wouldn't be too hard to make a `DefaultDebugWriter` like there is a `DefaultErrorWriter` or indeed change it over to a function call instead of using an io.Writer.... I can take a crack at this sometime soon.

**Comment by sneak** [comment]
Actually, it would appear that zerolog (my use case) already supports being used as an io.Writer, so it might be a pretty minimal change to gin.  I'll work on this tomorrow.

**Comment by sneak** [comment]
Turns out I was reading `master` - this is already possible by just overwriting `gin.DefaultWriter`, whereas the current released version still uses a hardcoded `os.Stderr`.  I was able to fix it by overwriting `gin.DefaultWriter` and `gin.DefaultErrorWriter` in latest `master`, i.e.: `go get github.com/gin-gonic/gin@db9174ae0c2587fe1c755def0f88cb9aba9e9641`.

**Comment by sneak** [comment]
This was fixed in #1891 and is on master.

**Comment by fatalc** [comment]
[debug.go#L24](https://github.com/gin-gonic/gin/blob/1d0f938f28d7e592623264591605dc22ac141cc5/debug.go#L24)
```go
var DebugPrintRouteFunc func(httpMethod, absolutePath, handlerName string, nuHandlers int)
```

`DebugPrintRouteFunc` now using for custom debug logging output if set.
